add memory hotplug design proposal (#230)
kubevirt/community: design-proposals/memory-hotplug.md

That number is needed by LibVirt and will default to 4 times the initial number of sockets if not set. That default
value will be configurable in the KubeVirt CR. On VM startup, that value will translate to a new `maxGuest` entry
under `spec.domain.memory` in the VMI object. However, all VMI Memory fields will stay immutable, and Memory guest will
only be adjustable on the VM object. Once the hoplug action begins, subsequent changes to the number of guest would be
rejected until the hotplug process completes. Memory Resource update on the pod level must occur before the actual
hotplug action on the Libvirt level, hence, during the hotplug action a notifitcation will be created upon successful
update of the Memory resources on the pod level. It will indicate that now it safe to proceed to the hotplug action on
the Libvirt/QEMU level.

## Drawbacks / Linmitations

* Workload update method should be defined explicitly as `VMLiveUpdateFeatures` under the Kubevirt configuration.
* This feature will be incompatible with memory requests/limits, at least initially, since those values need to
  dynamically scale according to the current number of enabled memory

> * Set the memory increment (new_guest - old_guest) for online memory expansion An integer multiple of 128 MB (operating

Author: Can you elaborate on this point?

> * Set the memory increment (new_guest - old_guest) for online memory expansion An integer multiple of 128 MB (operating

Reviewer: virtio-mem lets you hot(un)plug in blocks of 4MB, we don't need to align to 128MB.

> * Set the memory increment (new_guest - old_guest) for online memory expansion An integer multiple of 128 MB (operating

Author: > virtio-mem lets you hot(un)plug in blocks of 4MB, we don't need to align to 128MB.

oh... I will remove it

  system requirements)

> * Memory hotplug depends on numa, VMs will have to be `NUMAable` to enable Memory hotplug

Author: what I mean is that feature gates turns on NUMA because the configuration of memory depends on the numa of the worker node

> * Memory hotplug depends on numa, VMs will have to be `NUMAable` to enable Memory hotplug

Reviewer: I'm still not convinced I understood, every VM is `numable` in the sense that even with a single socket you can define a NUMA topology.

> * Memory hotplug depends on numa, VMs will have to be `NUMAable` to enable Memory hotplug

Author: looks like don't need this ,  i will remove it

#### API Examples

##### VM Spec API

```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  labels:
    kubevirt.io/vm: vm-cirros
  name: vm-cirros
spec:
  running: false
  liveUpdateFeatures:
  +   memory
+     maximumGuest: 4Gi
  template:
    metadata:
      labels:
        kubevirt.io/vm: vm-cirros
    spec:
      domain:
        memory:
          guest: 4Gi
```

##### VMI Spec API

```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  labels:
    special: vm-cirros
  name: vm-cirros
spec:
  domain:
    memory:
+     maxGuest: 10Gi
  guest: 8Gi

```

## Status API

> The VMI status must indicate whether Memory hotplug is taking place. As the hotplug process begins the current state

Reviewer: Can you reword this?

The VM/VMI will have a new field `status.memory` that looks like the following:

```yaml
status:
  memory:
    guestAtBoot: 4Gi
    guestCurrent: 8Gi
```

It's pretty self-explanatory but `guestAtBoot` shows how much memory the VM booted with and `guestCurrent` shows the currently available memory to the guest, so of course this last field changes when a hotplug is completed.

Author: I don't think you need to record what you want to change, you just need to have the original value。 `guestCurrent` is not necessary and can be represented by `spec.domain.memory.guest`

Reviewer: We do actually need `guestCurrent` as `spec.domain.memory.guest` can be different that `status.memory.guestCurrent` after a successful hot(un)plug operation as you're not always guaranteed to get all the memory you requested.

This is especially true when you're resizing down memory.

`guestCurrent` track how actual memory you have, `guest` track what the request was.

Author: Is your guestCurrent always equal to `spec.domain.memory.maxGuest` - `spec.domain.memory.guest` ?  If so, if this value is useful elsewhere, it can be calculated directly, and there is no need to add a field

Author: After reading the https://github.com/kubevirt/kubevirt/pull/10184 code, I understand what you mean. What I want to express is whether it can be implemented like cpu without caring about whether the operation is successful.   I think you only need to have one field in `status.memory` to be consistent with the domain, in fact, it is not very meaningful to record the actual memory size of the vm.  Can you give me an example if you feel it's important?

Reviewer: It is implemented like that, in https://github.com/kubevirt/kubevirt/pull/10184 if you request a memory hotplug KubeVirt will mark it as successful if you can hot(un)plug any memory at all, but we have to note somewhere what's the guest available memory as it really can be different from what requested, especially when resizing down memory.

Otherwise how can you tell how much memory your guest currently has? 

A simple example, you have a 4Gi VM, you want to hotplug another 4Gi of memory, for some reason the system can only give you 2 so you end up with a 6Gi VM and the hotplug is marked as successful. If you don't write in `status.memory.guestCurrent` how can you tell how much memory you have? That's a pretty important thing to know.

Author: Partial success when the requested resources are insufficient?

of `spec.domain.memory` needs to be propagated back to the VM/VMI `status.currentMemory` to allow the user to have an
indication of the hotplug processing.

```
	// Memory allow specified the detailed inside the vmi.
		CurrentMemory *Memory `json:"currentMemory,omitempty"`
```

Conditions on the VM and VMI should indicate the status of the pending Memory hotplug actions. When errors occur, the
conditions should propagate the errors to the VM.Status.Conditions field so users can receive feedback into the status
of their declared state.

```
// Indicates that the VMI is in progress of Hot memory Plug/UnPlug
VirtualMachineInstanceMemoryChange VirtualMachineInstanceConditionType = "HotMemoryChange"
```

Upon succesful hotplug action the condition will be removed. Otherwise, a human readable reason and message indicating
whether the hotplug change is pending or encountering errors.

## Extensibility

> We will be able to use the `VMLiveUpdateFeatures` section to manage things like cpu/memory hotplug .

Reviewer: You can remove this section, it was already mentioned in the CPU hotplug design where the `LiveUpdateFeatures` API was introduced.

Author: ok

## Update/Rollback Compatibility

The feautre doesn't intorduce any breaking changes so existing functionality of versions prior to this feature should
not be impacted upon update/rollback. However if existing workloads would like to opt-in for this feature after an
update, the VMI should be re-created since the addition of new fields to the VMI API.

## Functional Testing Approach

Functional tests will simply:

> * Create and start a VM with `memeory` set as one of th `liveUpdateFeatures`

Reviewer: ```suggestion
* Create and start a VM with `memory` set as one of th `liveUpdateFeatures`
```

> * Modify the number of memory guest for the VM

Reviewer: ```suggestion
* Modify the amount of guest memory for the VM
```

> * Ensure the guest sees the changed number of memory

Reviewer: ```suggestion
* Ensure that additional guest memory gets hotplugged, possibly matching the request. This is visible in `status.memory.guestCurrent`
```

* Ensure the memory resources on the virt-launcher pod were adjusted accordingly

> * Ensure interoperability with non-hotplug related live-migration occasions (workload-update, evacuation,

Reviewer: Can you elaborate on this point?

Author: Because the current implementation is dependent on migration, you need to consider some of the scenarios involved in migration

Reviewer: Can you do an example?

Author: > Can you give an example?

Sorry, I want to say cpu hotplug is dependent on migration.

  user-triggered)

## Implementation Phases

### Phase 1: Creating the KubeVirt API and linking it to the LibVirt memory hotplug API

> ### Phase 2: Triggering a live migration when the number of sockets changes, to adjust the virt-launcher memory resources

Reviewer: ```suggestion
### Phase 2: Triggering a live migration when guest memory changes, to adjust the virt-launcher memory resources
```

The
VMI [workload update controller](https://github.com/kubevirt/kubevirt/blob/ce8aab7874d2e1586787e3e2a17306b7edca1b8a/pkg/virt-controller/watch/workload-updater/workload-updater.go#L476)
is extended to live migrate VMIs to satisfy updates to a VMI resource request/limits field.

Sequence of events:

* User updates memory guest on a vm.spec.template.spec.domain.memory

> * VM controller determines the request/limits required and writes the necessary changes to the active vmi.Spec

Reviewer: ```suggestion
* VM controller determines the required memory requests and writes the necessary changes to the active vmi.Spec
```

> * VM controller writes the required request/limits in the vmi.Status

Reviewer: No need to write anything in the vmi.Status about reqs/limits. I think you can remove this.

Author: update from request/limits to request ?

Reviewer: Not in the status section, there's no need. Requests are already tracked in the spec.

Author: In fact, the 'vmi.Status' here is what you say' vmi.status.memory '

Author: enenen... I think cpu and memory  hot(un)plug should be aligned.     `status.memory.currentGuest` ==  `spec.domain.memory`

Reviewer: That's not always possible, memory hotplug doesn't work like CPU hotplug, you may never get the full amount of memory you requested with virtio-mem.

> * VMI workload update controller detects the mismatch between the resource request/limits on the pod vs vmi.Spec and

Reviewer: Can you rewrite this section? Most things are correct but the actual sequence is described in this commit https://github.com/kubevirt/kubevirt/pull/10184/commits/d0350f2e7bb1af2d5834d1fc146a514dde18e786.

  live migrates the VMI to satisfy vmi.spec
* Migration controller constructs a new target pod with the desired memory request/limits
* Migration controller identfies a match between the required request/limits on the vmi.Status and those of the VMI
  target pod
* Migration controller marks the VMI as safe to hotplug.
* VMI live migrates.
* virt-handler identifies the mark and triggers the hotplug/hot-unplug LibVirt action after the live migration completes
* VM status is updated to reflect that the pending hotplug/hot-unplug actions have completed

### Phase 3 (future): Using the [Kubernetes vertical scaling API](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources) once it reaches a more mature state

The inplace vertical pod feature allows us to dynamically change the VMI pods resource request/limits. This feature will
allow us to avoid live migration, however the issue for us is timing. This feature is alpha in Kubernetes 1.27. We need
it to be at least in beta to integrate it without risking stability and API reliability.

> Sequence of events:

Reviewer: I would remove the sequence of events here, it's going to be exactly the same as the previous section minus live migration, also it's more related to the implementation itself.

Author: remove it
